### ISSUE TITLE
Inconsistent behavior of gain changes

### ISSUE THREAD

**Issue by rtoy** [WebAudio/web-audio-api#1845]
See #1842, in particular https://github.com/WebAudio/web-audio-api/issues/1842#issuecomment-476878260

The issue is:

> The next test case is changing gain in the absence of a signal. The test creates an oscillator and a gain node, hooks them up without starting the oscillator, then calls setTargetAtTime on the gain node, then sets up a delayed call to start the oscillator in 1 second. It sets up one more delayed call to read the gain value 1 second later. In Chrome and Firefox, no sound is heard when the oscillator is started, which I would interpret to mean that the gain changes right away, except that Chrome reports a gain value of 1 at this point in time, and 0 a second later when the gain value is re-read. Firefox always reports a gain value of 1, but I don't think Firefox supports reading of gain values (which is another difference in behavior).

> In Safari and Edge, a sound is heard when the oscillator starts, and the gain values that are output seem to indicate that the movement of the gain value towards the target doesn't start until a signal is detected at the gain node.

> Again, I'm not sure which of these is out of spec, but for my purposes, the behavior of Chrome and Firefox is what I would have expected in the sense that no sound is heard.

> A live link is available at https://phet-dev.colorado.edu/jb-tests/gain-change-without-signal-tester.html and the source code is at https://github.com/phetsims/tambo/blob/master/html/gain-change-without-signal-tester.html.

**Comment by rtoy** [comment]
A quick summary of what I think the test is doing:

```
gainNode = new GainNode(context);
gainNode.gain.setValueAtTime(0, context.currentTime);
gainNode.connect(context.destination);
osc = new OscillatorNode(context, {type: "triangle"});
osc.frequency.setValueAtTime(220, context.currentTime);
osc.connect(gainNode);
gainNode.gain.setTargetAtTime(0, context.currentTime, 0.1);

// Wait one second
osc.start()
```

Because the time constant is 0.1, I do not expect to hear anything 1 sec after the ramp starts.  (Well, there is a tone, but it's very, very faint.)  I think Safari and Edge are wrong if you hear a full-amplitude tone fading to 0.

The value for the getter is issue #1788, which is being clarified now.

That chrome returns 1 when the oscillator starts (but 1 sec after the automation starts) is a bug in Chrome.  Firefox has a known issue that it doesn't return the value from an automation.  This is what #1788 is meant to specify.

**Comment by rtoy** [comment]
I tried Edge and I do hear a full amplitude tone fading to 0.  I think I know the problem there because I think Chrome had the same issue.  As an optimization, if the input is known to be silent, the output is marked as silent.  This used to bypass any processing of the AudioParam.  When the oscillator starts, the AudioParam starts processing, so you hear the tone fading out.

**Comment by rtoy** [comment]
This appears to be caused by bugs in the browsers and is not a bug in the spec.

@jbphet  WDYT?

**Comment by jbphet** [comment]
> @jbphet WDYT?

I don't know the spec well enough to comment intelligently on whether the problem lies in an ambiguity in the spec or the behavior of the browsers.  Essentially, that's what I'm trying to figure out so that I can, in my own small way, help to get the behavior consistent across browsers.  My main motivation is the the inconsistencies across browsers is requiring workaround and browser-specific code in my web apps, which I - and I think anyone else developing web apps that make use of sound - would like to avoid.

**Comment by rtoy** [comment]
Ok.  Spent more time looking into this on Chrome.  The current output after pressing "Go":
```
created and connected gain node
created and connected oscillator (did not start)
gainNode.gain.value = 1
oscillator will start in 1000 ms
started oscillator
when oscillator started, gainNode.gain.value = 1
after delay, gainNode.gain.value = 0
```
The first value for gainNode.gain is potentially correct.  Depending on where the audio thread is, reading the value right after `setValueAtTime` can produce 0 or 1.

The second value is 1, which is incorrect.  Sort of.  This is a consequence of Chrome's autoplay policy.  The context is stopped until you press "Go" to create a user-gesture, and when you call `start()`, the context is resumed.  So, no processing happened and the gain value is 1, but could be 0, depending on timing of the audio thread.

The last value is correctly 0.

I modified your code so that the function `clearLog` does `audioContext.resume()` so that the context is resumed with a user-gesture.  Then the second gain value is correctly printed as 0.

(You can see that autoplay is in effect if you open the dev console after loading the page and looking at `audioContext`.  The `currentTime` is 0 and stays that way.)

**Comment by rtoy** [comment]
I think the spec is clear on how this should behave (or will be once #1788 is fixed), and the autoplay policy is accounted for  The printed values should be 0, except, perhaps the first one which could be 0 or 1.

It's a bug if an implementation does something different, I think.

**Comment by rtoy** [comment]
Closing since the spec is clear on what should happen.  There should be no output at all.  The values printed in the log will get fixed in #1788
